Add a section on the git history to git.rst (#3064)
hyperspy/hyperspy: doc/dev_guide/git.rst

   $ git rebase --onto RELEASE_next_patch RELEASE_next_minor bug_fix_branch

> Cleaning the commit history
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
Keeping the git history clean
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

> For review, and for revisiting changes at a later point, it is advisable to clean the git history when a pull request is split up in too many commits. Commits that only fix formatting issues or failing tests usually can be *squashed* (merged) into the previous commits. When using a GUI for interaction with *git*, check out its features for joining and reordering commits.

Reviewer: ```suggestion
For review, and for revisiting changes at a later point, it is advisable to keep a "clean" git history, i.e. a meaningful succession of commits. In some cases, it is useful to rewrite the git history to keep it more readable:
- it is not always possible to keep a clean history and quite often the code development follows an exploratory process with code changes going back and forth, etc. 
- Commits that only fix typographic mistakes, formatting or failing tests usually can be *squashed* (merged) into the previous commits.

When using a GUI for interaction with *git*, check out its features for joining and reordering commits.
```

When using git in the command line, use ``git rebase`` with the *interactive* option. For example to rearrange the last five commits:

.. code:: bash

  $ git rebase -i HEAD~5

In a text editor, you can then edit the commit history. If you have commits ``a...e`` and want to merge ``b`` and ``e`` into ``a`` and ``d``, respectively, while moving ``c`` to the end of the hisotry, your file would look the following:

.. code::  text

  pick a ...
  squash b ...
  pick d ...
  squash e ...
  pick c ...

Afterwards you get a chance to edit the commit messages.

Finally, to push the changes use a ``+`` in front of the branch name, to override commits you have already pushed to github previously:

.. code:: bash

  git push origin +lumberjack-branch

> See for example `Git good - The magic of keeping a clean git history <https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/>`_ for a more detailed blog post on this subject.

Reviewer: This links doesn't have https and is flagged as insecure; maybe uses this link instead:
https://about.gitlab.com/blog/2018/06/07/keeping-git-commit-history-clean/?

Author: Very nice, it is more concise and includes some examples as well.

> See for example `Git good - The magic of keeping a clean git history <https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/>`_ for a more detailed blog post on this subject.

Author: ```suggestion
See for example `How (and why!) to keep your Git commit history clean <https://about.gitlab.com/blog/2018/06/07/keeping-git-commit-history-clean/>`_ for a more detailed blog post on this subject.
```
